lua/rainbow-delimiters/strategy/global.lua

```diff
@@ -123,30 +158,41 @@ local function setup_parser(bufnr, parser)
 
 		p:register_cbs {
 			on_changedtree = function(changes, tree)
-				if #changes == 0 then
-					-- Hack: an empty change set results from an undo.  Force a
-					-- full update by setting the maximum possible range
-					table.insert(changes, {tree:root():range()})
-				end
 				log.trace('Changed tree in buffer %d with languages %s', bufnr, lang)
 				-- HACK: As of Neovim v0.9.1 there is no way of unregistering a
 				-- callback, so we use this check to abort
 				if not lib.buffers[bufnr] then return end
 
-				-- Normalize the changes objects
-				changes = vim.tbl_map(normalize_change, changes)
+				-- if we have no parent language, then we use changes,
+				-- otherwise we use the whole tree's range
+				if p._parent_lang then
+					-- HACK: changes can accidentally overwrite highlighting in
+					-- injected code blocks
+
+					-- We have a parent language, so we are in an injected
+					-- language code block, thus we update all of the current
+					-- code block
+					changes = {{tree:root():range()}}
+				else
```

Comment:
Can you please elaborate on why you are using the full tree here?

Let's say I have Lua code with embedded Vim script:

```lua
print('Hello from lua')
vim.cmd [[
echo 'Hello from Vim'
echo 'tests'
echo 2 + 3 == 5
echo 'Goodbye from Vim'
]]
print('Goodbye from lua')
```

What is the scenario this is trying to solve?

Reply:
The scenario this tries to solve doesn't seem to show up in Lua code with embedded Vim script (presumably the treesitter implementation catching the changes is better there), but it does show up in markdown with injected languages.

E.g., if you have something like (but with a proper injected language block -- I am not sure how to nest them in markdown):

```markdown
Some markdown. 
`vim
echo 'Hello from Vim'
echo 'tests'
`
More markdown.
```
Then if you move the top line out of the injected language block at the top and move it back down, you get correct looking changes for the `vim` block, but the `markdown` changes are registered as the whole `vim` block changing. Thus Vim highlighting will be reset in the block (by the markdown changes), and then only updated on the first line (by the `vim` block changes). 

This is a pretty rare problem, but it show up in my testing with markdown.

Reply:
Does it still happen with the current code? I tried to replicate your example; first I executed `:move 1` while my cursor was on line 3, then I executed `:move 3` while my cursor was on line 2 to move the text back into the block. The result looks exactly as before. The changes the firs time around are

- markdown `{ { 0, 14, 1, 21 } }`
- vim `{}`

The second time around (moving the line back) the changes are

- markdown `{ { 0, 14, 1, 0 }, { 2, 0, 4, 0 } }`
- vim `{ { 2, 0, 3, 0 } }`

Could it be that the order is the other way around for you? I am running Neovim 0.9.4 on GNU/Linux.

Reply:
Here is the behavior for me with the current code:

https://github.com/HiPhish/rainbow-delimiters.nvim/assets/7075380/1b8025d8-9b77-47e4-996d-aa0acd279736

and with the entire `if`/`else` block commented out but 
```lua
changes = vim.tbl_map(normalize_change, changes)
```
added outside the `if`/`else` block instead:


https://github.com/HiPhish/rainbow-delimiters.nvim/assets/7075380/66b84842-dab0-4c25-9e54-1773ccb0aa9d


It might be because we move text differently? I move text with these keymaps:
```lua
-- keymaps for moving marked text up or down
vim.keymap.set("v", "J", ":m '>+1<CR>gv=gv", { silent = true, desc = "Move marked text down" })
vim.keymap.set("v", "K", ":m '<-2<CR>gv=gv", { silent = true, desc = "Move marked text up" })
```

Reply:
See my comment below. I think it might actually just be the different files behaving differently and not the commands.